enhancements/ingress/customize-error-code-pages.md

```diff
@@ -0,0 +1,121 @@
+exist, we should return 404 or something different as it otherwise may confuse end-customers.
+
+### Goals
+1. Enable the cluster administrator to specify custom error code pages for haproxy router to serve them at the right moment.
+
+### Non-Goal
+TBD
+
+## Proposal
+0. User creates a custom error code page configmap `http503page` in `openshift-config`
+1. The IngressController API is extended by adding an optional
+`HttpErrorCodePage` field with type string
+```go
+//This used to sync the configmap having customizable error code page created by cluster admin in openshift-config namespace.
+//to openshift-ingress which will be used by the router to serve appropriate error code pages.
+//If this field is empty, the default values i.e default shipped error code pages will be served by the haproxy router.
+HttpErrorCodePage string `json:"httpErrorCodePage,omitempty"`
+
+```
+Using the default, the router uses the default error pages used by haproxy.
+The `HttpErrorCodePage` field enables the cluster administrator to
+specify custom error code pages.
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpErrorCodePage: "http503page"  // http503page is the config map name with custom 
+                                    //503 error page created in openshift-config by cluster-admin. 
+```
+2. `errorfile` stanzas need to be added as per https://www.haproxy.com/blog/serve-dynamic-custom-error-pages-with-haproxy/
+3. default error pages will be added in openshift/router for the haproxy.
+4. A  new controller to sync the cluster admin created configmap having custom error pages from `openshift-config` 
+   to `openshift-ingress`.
+5. changes to ingress controller's function `desiredRouterDeployment` present in `deployment.go`.
+   of openshift/cluster-ingress-operator to mount the custom error code page configmap into the router deployment.
+
+
+### Validation
+Omitting `spec.httpErrorCodePage`  specifies the default behavior i.e serves default error code pages.
+If user specifies an invalid value, then the IngressController ignores the annotation.
+
+### User Stories
+
+#### As a cluster administrator, I expect to have custom error code pages getting served
+1. The administrator creates a configmap named http503page in the openshift-config namespace.
+2. The administrator patches the ingresscontroller to reference the http503page configmap by name.
+3. The ingress operator copies the http503page configmap from the openshift-config namespace to the
+   openshift-ingress namespace.
+4. The ingress operator configures the router deployment with a volume and volume mount that mounts
+   the custom error pages from the http503page configmap on top of the existing error pages in the 
+   router pods.
+   
+### Risks and Mitigations
+No risk identified. This is a cosmetic change.
```

Comment:
I think there are a couple risks worth mentioning.

1) What if a user supplies an incomplete or confusing error page(s)? Or what if a user provides an absolutely massive HTML file for a trivial 503? Could that effect performance somehow?
2) Are there any applications that use the default ingress controller that expect a 503 in non-503 cases? Does the administrator need to exercise caution when creating custom error pages for the default ingress controller? (I would hope not, but it might be worth double checking)

Reply:
If the user does not provide proper file name in configmap then the custom error page will not be served. The default shall be served in that case

Yeah admin will need to take some precaution. I will list them. May be we can do some validations at our end.